ENVIRONMENT.md
# Environment Variables

A list of environment variables that are used by the ar.io node and its
components. These can be found in the [`src/config.ts`][src/config.ts] file.

## Server Configuration

| Variable                      | Description                                | Default      |
| ----------------------------- | ------------------------------------------ | ------------ |
| PORT                          | HTTP server port                           | 4000         |
| ADMIN_API_KEY                 | API key for accessing admin HTTP endpoints | Random value |
| ADMIN_API_KEY_FILE            | File containing admin API key              | -            |
| HEALTHCHECK_INTERVAL_SECONDS  | Interval for healthcheck                   | 60           |
| HEALTHCHECK_TIMEOUT_SECONDS   | Timeout for healthcheck                    | 10           |
| HEALTHCHECK_FAILURE_THRESHOLD | Failure threshold for healthcheck          | 3            |
| HEALTHCHECK_SUCCESS_THRESHOLD | Success threshold for healthcheck          | 2            |

## AWS Configuration

| Variable                      | Description                   | Default |
| ----------------------------- | ----------------------------- | ------- |
| AWS_ACCESS_KEY_ID             | AWS access key ID             | -       |
| AWS_SECRET_ACCESS_KEY         | AWS secret access key         | -       |
| AWS_REGION                    | AWS region                    | -       |
| AWS_ENDPOINT                  | AWS endpoint URL              | -       |
| AWS_S3_CONTIGUOUS_DATA_BUCKET | S3 bucket for contiguous data | -       |
| AWS_S3_CONTIGUOUS_DATA_PREFIX | S3 prefix for contiguous data | -       |

## AR.IO Network

| Variable           | Description        | Default |
| ------------------ | ------------------ | ------- |
| AR_IO_WALLET       | AR.IO wallet       | -       |
| IO_PROCESS_ID      | AR.IO process ID   | -       |
| AR_IO_NODE_RELEASE | AR.IO node release | -       |

## AO Configuration

| Variable       | Description             | Default |
| -------------- | ----------------------- | ------- |
| AO_MU_URL      | AO Memory Unit URL      | -       |
| AO_CU_URL      | AO Compute Unit URL     | -       |
| AO_GRAPHQL_URL | AO GraphQL endpoint URL | -       |
| AO_GATEWAY_URL | AO Gateway URL          | -       |

## ARNS Configuration

| Variable                                                  | Description                                            | Default           |
| --------------------------------------------------------- | ------------------------------------------------------ | ----------------- |
| ARNS_ROOT_HOST                                            | Root host for ARNS                                     | -                 |
| ARNS_CACHE_TYPE                                           | Cache type for ARNS resolution data                    | lmdb              |
| ARNS_CACHE_MAX_KEYS                                       | Maximum number of keys to store in cache               | 10000             |
| ARNS_CACHE_TTL_SECONDS                                    | TTL for resolved data in cache                         | 3600              |
| ARNS_NAMES_CACHE_TTL_SECONDS                              | TTL for base names in cache                            | 300               |
| ARNS_NAME_LIST_CACHE_MISS_REFRESH_INTERVAL_SECONDS        | Refresh interval of base ArNS name cache on cache miss | 10                |
| ARNS_NAME_LIST_CACHE_HIT_REFRESH_INTERVAL_SECONDS         | Refresh interval of base ArNS name cache on cache hit  | 3600              |
| ARNS_RESOLVER_OVERRIDE_TTL_SECONDS                        | Override ANT record TTL                                | 0                 |
| ARNS_RESOLVER_PRIORITY_ORDER                              | Resolution priority order                              | on-demand,gateway |
| TRUSTED_ARNS_GATEWAY_URL                                  | Trusted ArNS gateway URL                               | -                 |
| ARNS_ON_DEMAND_CIRCUIT_BREAKER_TIMEOUT_MS                 | AO CU response timeout                                 | 15000             |
| ARNS_ON_DEMAND_CIRCUIT_BREAKER_ERROR_THRESHOLD_PERCENTAGE | Circuit breaker error threshold                        | 50                |
| ARNS_ON_DEMAND_CIRCUIT_BREAKER_ROLLING_COUNT_TIMEOUT_MS   | Circuit breaker rolling count timeout                  | 60000             |
| ARNS_ON_DEMAND_CIRCUIT_BREAKER_RESET_TIMEOUT_MS           | Circuit breaker reset timeout                          | 300000            |
| ARNS_ANT_STATE_CACHE_MISS_REFRESH_INTERVAL_SECONDS        | Refresh interval of ANT state cache on cache miss      | 10                |
| ARNS_ANT_STATE_CACHE_HIT_REFRESH_INTERVAL_SECONDS         | Refresh interval of ANT state cache on cache hit       | 300               |

## Sandboxing

| Variable            | Description                          | Default |
| ------------------- | ------------------------------------ | ------- |
| SANDBOX_PROTOCOL    | Protocol for sandboxing redirects    | https   |
| SANDBOX_HOST        | Host for sandboxing redirects        | -       |
| SANDBOX_PORT        | Port for sandboxing redirects        | -       |
| SANDBOX_PATH_PREFIX | Path prefix for sandboxing redirects | -       |

## Arweave Nodes

| Variable                               | Description                                | Default |
| -------------------------------------- | ------------------------------------------ | ------- |
| TRUSTED_NODE_URL                       | Trusted Arweave node URL                   | -       |
| TRUSTED_GATEWAY_URL                    | Trusted gateway URL                        | -       |
| TRUSTED_GATEWAYS_URLS                  | Trusted gateways URLs and weights          | -       |
| CHUNK_POST_URLS                        | Chunk POST URLs                            | -       |
| SECONDARY_CHUNK_POST_URLS              | Secondary chunk POST URLs                  | -       |
| SECONDARY_CHUNK_POST_CONCURRENCY_LIMIT | Secondary chunk POST concurrency limit     | -       |
| SECONDARY_CHUNK_POST_MIN_SUCCESS_COUNT | Secondary chunk POST minimum success count | -       |
| CHUNK_POST_RESPONSE_TIMEOUT_MS         | Chunk POST response timeout                | -       |
| CHUNK_POST_ABORT_TIMEOUT_MS            | Chunk POST abort timeout                   | -       |

## Header Cache

| Variable         | Description               | Default |
| ---------------- | ------------------------- | ------- |
| CHAIN_CACHE_TYPE | Cache type for chain data | lmdb    |
| REDIS_CACHE_URL  | Redis URL                 | -       |
| REDIS_USE_TLS    | Whether to use TLS        | false   |

## Data Retrieval

| Variable                   | Description                         | Default |
| -------------------------- | ----------------------------------- | ------- |
| ON_DEMAND_RETRIEVAL_ORDER  | Order for on-demand data retrieval  | -       |
| BACKGROUND_RETRIEVAL_ORDER | Order for background data retrieval | -       |

## Contiguous Data Cache

| Variable                                | Description                                 | Default |
| --------------------------------------- | ------------------------------------------- | ------- |
| CONTIGUOUS_DATA_CACHE_CLEANUP_THRESHOLD | Threshold for contiguous data cache cleanup | -       |

## Data Verification

| Variable                                      | Description                               | Default |
| --------------------------------------------- | ----------------------------------------- | ------- |
| ENABLE_BACKGROUND_DATA_VERIFICATION           | Enable background data verification       | false   |
| BACKGROUND_DATA_VERIFICATION_INTERVAL_SECONDS | Interval for background data verification | 600     |

## Filesystem Cleanup

| Variable                                 | Description                                                                 | Default |
| ---------------------------------------- | --------------------------------------------------------------------------- | ------- |
| FS_CLEANUP_WORKER_BATCH_SIZE             | Number of files to process in each batch                                    | 2000    |
| FS_CLEANUP_WORKER_BATCH_PAUSE_DURATION   | Pause duration between batches in milliseconds                              | 5000    |

CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).

## [Unreleased]

## [Release 23] - 2025-01-13

### Added

- Added `FS_CLEANUP_WORKER_BATCH_SIZE`,
  `FS_CLEANUP_WORKER_BATCH_PAUSE_DURATION`, and
  `FS_CLEANUP_WORKER_RESTART_PAUSE_DURATION` environment variables to allow
  configuration of number of contiguous data files cleaned up per batch, the
  pause between each batch, and the pause before restarting the entire cleanup
  process again.
- Added `data_items_unbundled_total` Prometheus metric that counts the total
  number of data items unbundled, including those that did not match the
  unbundling filter.
- Added a `parent_type` label that can be one of `transaction` or `data_item`
  to data item indexing metrics.
- Added a `files_cleaned_total` total Prometheus metric to enable monitoring of
  contiguous data cleanup.
- Added support for specifying the admin API via a file specified by the
  `ADMIN_API_KEY_FILE` environment variable.
- Added experimental support for posting chunks in a non-blocking way to
  secondary nodes specified via a comma separate list in the
  `SECONDARY_CHUNK_POST_URLS` environment variable.

### Changed

- Renamed the `parent_type` lable to `contiguous_data_type` on bundle metrics
  to more accurately reflect the meaning of the label.
- Reduced the maximum time to refresh the ArNS name list to 10 seconds to
  minimize delays in ArNS availability after a new name is registered.
- Changed `/ar-io/admin/queue-bundle` to wait for `bundles` rows to be written
  to the DB before responding to ensure that errors that occur due to DB
  contention are not silently ignored.
- Data items are now flushed even when block indexing is stopped. This allows
  for indexing batches of data items using the admin API with block indexing
  disabled.
- Adjust services in `docker-compose` to use `unless-stopped` as their restart
  policy. This guards against missing restarts in the case where service
  containers exit with a success status even when they shouldn't.

### Fixed

- Added missing `created_at` field in `blocked_names` table.
- Fixed broken ArNS undername resolution.

## [Release 22] - 2024-12-18

### Added

- Added the ability to block and unblock ArNS names (e.g., to comply with
  hosting provider TOS). To block a name, POST `{ "name": "<name to block>" }`
  to `/ar-io/admin/block-name`. To unblock a name, POST
  `{ "name": "<name to unblock>" }` to `/ar-io/admin/unblock-name`.

### Changed

- Return an HTTP 429 response to POSTs to `/ar-io/admin/queue-bundle` when the
  bundle data import queue is full so that scripts queuing bundles can wait
  rather than overflowing it.

### Fixed

- Adjust ArNS length limit from <= 48 to <= 51 to match the limit enforced by
  the AO process.

## [Release 21] - 2024-12-05

### Added

- Added a ClickHouse auto-import service. When enabled, it calls the Parquet
  export API, imports the exported Parquet into ClickHouse, moves the Parquet
  files to an `imported` subdirectory, and deletes data items in SQLite up to
  where the Parquet export ended. To use it, run Docker Compose with the
  `clickhouse` profile, set the `CLICKHOUSE_URL` to `http://clickhouse:8123`,
  and ensure you have set an `ADMIN_KEY`.
  Using this configuration, the core service will also combine results from
  ClickHouse and SQLite when querying transaction data via GraphQL. Note: if
  you have a large number of data items in SQLite, the first export and
  subsequent delete may take an extended period. Also, this functionality is
  considered **experimental**. We expect there are still bugs to be found in it
  and we may make breaking changes to the ClickHouse schema in the future. If
  you choose to use it in production (not yet recommended), we suggest backing
  up copies of the Parquet files found in `data/parquet/imported` so that they
  can be reimported if anything goes wrong or future changes require it.
- Added a background data verification process that will attempt to recompute
  data roots for bundles and compare them to data roots indexed from Arweave

src/config.ts
/**
 * AR.IO Gateway
 * Copyright (C) 2022-2023 Permanent Data Solutions, Inc. All Rights Reserved.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
import { canonicalize } from 'json-canonicalize';
import { isMainThread } from 'node:worker_threads';
import { existsSync, readFileSync } from 'node:fs';

import { createFilter } from './filters.js';
import * as env from './lib/env.js';
import { release } from './version.js';

//
// HTTP server
//

// HTTP server port
export const PORT = +env.varOrDefault('PORT', '4000');

// API key for accessing admin HTTP endpoints
// It's set once in the main thread
export let ADMIN_API_KEY = isMainThread
  ? env.varOrRandom('ADMIN_API_KEY')
  : undefined;

const ADMIN_API_KEY_FILE = isMainThread
  ? env.varOrUndefined('ADMIN_API_KEY_FILE')
  : undefined;

if (ADMIN_API_KEY_FILE !== undefined) {
  if (!existsSync(ADMIN_API_KEY_FILE)) {
    throw new Error(`ADMIN_API_KEY_FILE not found: ${ADMIN_API_KEY_FILE}`);
  }
  ADMIN_API_KEY = readFileSync(ADMIN_API_KEY_FILE).toString().trim();
}

//
// Nodes
//

// Trusted Arweave node URL (for syncing the chain and retrieving chunks)
export const TRUSTED_NODE_URL = env.varOrDefault(
  'TRUSTED_NODE_URL',
  'https://arweave.net',
);

// Trusted gateway URL (for retrieving contiguous data)
export const TRUSTED_GATEWAY_URL = env.varOrDefault(
  'TRUSTED_GATEWAY_URL',
  'https://arweave.net',
);

// Trusted gateway URLs (for retrieving contiguous data)
export const TRUSTED_GATEWAYS_URLS = JSON.parse(
  env.varOrDefault(
    'TRUSTED_GATEWAYS_URLS',
    JSON.stringify({
      [TRUSTED_GATEWAY_URL]: 1,
    }),
  ),
) as Record<string, number>;

// Validate URLs and weights
Object.entries(TRUSTED_GATEWAYS_URLS).forEach(([url, weight]) => {
  try {
    new URL(url);
  } catch (error) {
    throw new Error(`Invalid URL in TRUSTED_GATEWAYS_URLS: ${url}`);
  }
  if (typeof weight !== 'number' || weight <= 0) {
    throw new Error(
      `Invalid weight in TRUSTED_GATEWAYS_URLS for ${url}: ${weight}`,
    );
  }
});

export const TRUSTED_GATEWAYS_REQUEST_TIMEOUT_MS = +env.varOrDefault(
  'TRUSTED_GATEWAYS_REQUEST_TIMEOUT_MS',
  '10000',
);

export const WEIGHTED_PEERS_TEMPERATURE_DELTA = +env.varOrDefault(
  'WEIGHTED_PEERS_TEMPERATURE_DELTA',
  '0.1',
);

export const ARWEAVE_NODE_IGNORE_URLS: string[] =
  env.varOrUndefined('ARWEAVE_NODE_IGNORE_URLS')?.split(',') ?? [];

// Trusted chunk POST URLs (for posting chunks received at /chunk)
export const CHUNK_POST_URLS = env
  .varOrDefault('CHUNK_POST_URLS', `${TRUSTED_NODE_URL}/chunk`)
  .split(',');

const SECONDARY_CHUNK_POST_URLS_STRING = env.varOrUndefined(
  'SECONDARY_CHUNK_POST_URLS',
);
export const SECONDARY_CHUNK_POST_URLS =
  SECONDARY_CHUNK_POST_URLS_STRING !== undefined
    ? SECONDARY_CHUNK_POST_URLS_STRING.split(',')
    : [];

const SECONDARY_CHUNK_POST_CONCURRENCY_LIMIT_STRING = env.varOrUndefined(
  'SECONDARY_CHUNK_POST_CONCURRENCY_LIMIT',
);
export const SECONDARY_CHUNK_POST_CONCURRENCY_LIMIT =
  SECONDARY_CHUNK_POST_CONCURRENCY_LIMIT_STRING !== undefined
    ? +SECONDARY_CHUNK_POST_CONCURRENCY_LIMIT_STRING
    : 2;

const SECONDARY_CHUNK_POST_MIN_SUCCESS_COUNT_STRING = env.varOrUndefined(
  'SECONDARY_CHUNK_POST_MIN_SUCCESS_COUNT',
);
export const SECONDARY_CHUNK_POST_MIN_SUCCESS_COUNT =
  SECONDARY_CHUNK_POST_MIN_SUCCESS_COUNT_STRING !== undefined
    ? +SECONDARY_CHUNK_POST_MIN_SUCCESS_COUNT_STRING
    : 1;

// Chunk POST response timeout in milliseconds
const CHUNK_POST_RESPONSE_TIMEOUT_MS_STRING = env.varOrUndefined(
  'CHUNK_POST_RESPONSE_TIMEOUT_MS',
);
export const CHUNK_POST_RESPONSE_TIMEOUT_MS =
  CHUNK_POST_RESPONSE_TIMEOUT_MS_STRING !== undefined
    ? +CHUNK_POST_RESPONSE_TIMEOUT_MS_STRING
    : undefined;

// Chunk POST abort timeout in milliseconds
const CHUNK_POST_ABORT_TIMEOUT_MS_STRING = env.varOrUndefined(
  'CHUNK_POST_ABORT_TIMEOUT_MS',
);
export const CHUNK_POST_ABORT_TIMEOUT_MS =
  CHUNK_POST_ABORT_TIMEOUT_MS_STRING !== undefined
    ? +CHUNK_POST_ABORT_TIMEOUT_MS_STRING
    : undefined;

export const CHUNK_POST_MIN_SUCCESS_COUNT = +env.varOrDefault(
  'CHUNK_POST_MIN_SUCCESS_COUNT',
  '3',
);

//
// Data
//

// On-demand data retrieval priority order
export const ON_DEMAND_RETRIEVAL_ORDER = env
  .varOrDefault(
    'ON_DEMAND_RETRIEVAL_ORDER',
    's3,trusted-gateways,chunks,tx-data',
  )
  .split(',');

// Background data retrieval priority order
export const BACKGROUND_RETRIEVAL_ORDER = env
  .varOrDefault(
    'BACKGROUND_RETRIEVAL_ORDER',
    'chunks,s3,trusted-gateways,tx-data',
  )
  .split(',');

//
// Indexing
//

// Whether or not to run indexing processes (used on readers when running with
// replication)
export const START_WRITERS =
  env.varOrDefault('START_WRITERS', 'true') === 'true';

// Indexing range
export const START_HEIGHT = +env.varOrDefault('START_HEIGHT', '0');
export const STOP_HEIGHT = +env.varOrDefault('STOP_HEIGHT', 'Infinity');

// Filter determining which ANS-104 bundles to unbundle
export const ANS104_UNBUNDLE_FILTER_PARSED = JSON.parse(
  env.varOrDefault('ANS104_UNBUNDLE_FILTER', '{"never": true}'),
);
export const ANS104_UNBUNDLE_FILTER_STRING = canonicalize(
  ANS104_UNBUNDLE_FILTER_PARSED,
);
export const ANS104_UNBUNDLE_FILTER = createFilter(
  JSON.parse(ANS104_UNBUNDLE_FILTER_STRING),
);

// Filter determining which ANS-104 data items to index
export const ANS104_INDEX_FILTER_PARSED = JSON.parse(
  env.varOrDefault('ANS104_INDEX_FILTER', '{"never": true}'),
);
export const ANS104_INDEX_FILTER_STRING = canonicalize(
  ANS104_INDEX_FILTER_PARSED,
);
export const ANS104_INDEX_FILTER = createFilter(
  JSON.parse(ANS104_INDEX_FILTER_STRING),
);

// The number of ANS-104 worker threads to run
export const ANS104_UNBUNDLE_WORKERS = +env.varOrDefault(
  'ANS104_UNBUNDLE_WORKERS',
  ANS104_UNBUNDLE_FILTER.constructor.name === 'NeverMatch' ? '0' : '1',
);

// The number of ANS-104 bundle downloads to attempt in parallel
export const ANS104_DOWNLOAD_WORKERS = +env.varOrDefault(
  'ANS104_DOWNLOAD_WORKERS',
  ANS104_UNBUNDLE_FILTER.constructor.name === 'NeverMatch' ? '0' : '5',
);

// Whether or not to attempt to rematch old bundles using the current filter
export const FILTER_CHANGE_REPROCESS =
  env.varOrDefault('FILTER_CHANGE_REPROCESS', 'false') === 'true';

// Whether or not to backfill bundle records (only needed for DBs that existed
// before unbundling was implemented)
export const BACKFILL_BUNDLE_RECORDS =
  env.varOrDefault('BACKFILL_BUNDLE_RECORDS', 'false') === 'true';

// Whether or not to write the data item signatures to the database
export const WRITE_ANS104_DATA_ITEM_DB_SIGNATURES =
  env.varOrDefault('WRITE_ANS104_DATA_ITEM_DB_SIGNATURES', 'false') === 'true';

// Whether or not to write the transaction signatures to the database
export const WRITE_TRANSACTION_DB_SIGNATURES =
  env.varOrDefault('WRITE_TRANSACTION_DB_SIGNATURES', 'false') === 'true';

// Whether or not to enable the data database WAL cleanup worker
export const ENABLE_DATA_DB_WAL_CLEANUP =
  env.varOrDefault('ENABLE_DATA_DB_WAL_CLEANUP', 'false') === 'true';

// The maximum number of data items to queue for indexing before skipping
// indexing new data items
export const MAX_DATA_ITEM_QUEUE_SIZE = +env.varOrDefault(
  'MAX_DATA_ITEM_QUEUE_SIZE',
  '100000',
);

// The maximum number of bundles to queue for unbundling before skipping
export const BUNDLE_DATA_IMPORTER_QUEUE_SIZE = +env.varOrDefault(
  'BUNDLE_DATA_IMPORTER_QUEUE_SIZE',
  '1000',
);

// The maximum number of data imports to queue for verification puroses
export const VERIFICATION_DATA_IMPORTER_QUEUE_SIZE = +env.varOrDefault(
  'VERIFICATION_DATA_IMPORTER_QUEUE_SIZE',
  '1000',
);

// The maximum number of data items indexed to flush stable data items
export const DATA_ITEM_FLUSH_COUNT_THRESHOLD = +env.varOrDefault(
  'DATA_ITEM_FLUSH_COUNT_THRESHOLD',
  '1000',
);

// The interval in seconds to flush stable data items
export const MAX_FLUSH_INTERVAL_SECONDS = +env.varOrDefault(
  'MAX_FLUSH_INTERVAL_SECONDS',
  '600',
);

export const BUNDLE_REPAIR_RETRY_INTERVAL_SECONDS = +env.varOrDefault(
  'BUNDLE_REPAIR_RETRY_INTERVAL_SECONDS',
  '300', // 5 minutes
);

export const BUNDLE_REPAIR_RETRY_BATCH_SIZE = +env.varOrDefault(
  'BUNDLE_REPAIR_RETRY_BATCH_SIZE',
  '1000',
);

//
// File system cleanup
//

// The number of files to process in each batch during cleanup
export const FS_CLEANUP_WORKER_BATCH_SIZE = +env.varOrDefault(
  'FS_CLEANUP_WORKER_BATCH_SIZE',
  '2000',
);

// The pause duration between cleanup batches in milliseconds
export const FS_CLEANUP_WORKER_BATCH_PAUSE_DURATION = +env.varOrDefault(
  'FS_CLEANUP_WORKER_BATCH_PAUSE_DURATION',
  '5000',
);

// The pause duration before restarting cleanup from the beginning in milliseconds
export const FS_CLEANUP_WORKER_RESTART_PAUSE_DURATION = +env.varOrDefault(
  'FS_CLEANUP_WORKER_RESTART_PAUSE_DURATION',
  `${1000 * 60 * 60 * 4}`, // every 4 hours
);

//
// Verification
//

// Whether or not to enable the background data verification worker
export const ENABLE_BACKGROUND_DATA_VERIFICATION =
  env.varOrDefault('ENABLE_BACKGROUND_DATA_VERIFICATION', 'false') === 'true';

export const BACKGROUND_DATA_VERIFICATION_INTERVAL_SECONDS = +env.varOrDefault(
  'BACKGROUND_DATA_VERIFICATION_INTERVAL_SECONDS',
  '600', // 10 minutes
);

export const BACKGROUND_DATA_VERIFICATION_WORKER_COUNT = +env.varOrDefault(
  'BACKGROUND_DATA_VERIFICATION_WORKER_COUNT',
  '1',
);

export const BACKGROUND_DATA_VERIFICATION_STREAM_TIMEOUT_MS = +env.varOrDefault(
  'BACKGROUND_DATA_VERIFICATION_STREAM_TIMEOUT_MS',
  `${1000 * 30}`, // 30 seconds
);

//
// GraphQL
//

export const TAG_SELECTIVITY = JSON.parse(
  env.varOrDefault(
    'TAG_SELECTIVITY',
    JSON.stringify({
      'Parent-Folder-Id': 20,
      Message: 20,
      'Drive-Id': 10,
      Process: 10,
      Recipient: 10,
      'App-Name': -10,
      'Content-Type': -10,
      'Data-Protocol': -10,
    }),
  ),
) as Record<string, number>;

// ClickHouse URL
export const CLICKHOUSE_URL = env.varOrUndefined('CLICKHOUSE_URL');

//
// Healthchecks
//

export const MAX_EXPECTED_DATA_ITEM_INDEXING_INTERVAL_SECONDS_STRING =
  env.varOrUndefined('MAX_EXPECTED_DATA_ITEM_INDEXING_INTERVAL_SECONDS');

export const MAX_EXPECTED_DATA_ITEM_INDEXING_INTERVAL_SECONDS =
  MAX_EXPECTED_DATA_ITEM_INDEXING_INTERVAL_SECONDS_STRING !== undefined
    ? +MAX_EXPECTED_DATA_ITEM_INDEXING_INTERVAL_SECONDS_STRING
    : undefined;

//
// ArNS and sandboxing
//

// The root host name to use for ArNS
export const ARNS_ROOT_HOST = env.varOrUndefined('ARNS_ROOT_HOST');
export const ROOT_HOST_SUBDOMAIN_LENGTH =
  ARNS_ROOT_HOST !== undefined ? ARNS_ROOT_HOST.split('.').length - 2 : 0;

// The protocol to use for sandboxing redirects (defaults to https)
export const SANDBOX_PROTOCOL = env.varOrUndefined('SANDBOX_PROTOCOL');

//
// AR.IO network
//

// The wallet for this gateway
export const AR_IO_WALLET = env.varOrUndefined('AR_IO_WALLET');

export const IO_PROCESS_ID = env.varOrDefault(
  'IO_PROCESS_ID',
  'agYcCFJtrMG6cqMuZfskIkFTGvUPddICmtQSBIoPdiA',
);

export const AR_IO_NODE_RELEASE = env.varOrDefault(
  'AR_IO_NODE_RELEASE',
  release,
);

//
// Header caching
//

// Cache type (lmdb, fs, or redis)
export const CHAIN_CACHE_TYPE = env.varOrDefault('CHAIN_CACHE_TYPE', 'lmdb');

// Redis URL
export const REDIS_CACHE_URL = env.varOrDefault(
  'REDIS_CACHE_URL',
  'redis://localhost:6379',
);

export const REDIS_USE_TLS =
  env.varOrDefault('REDIS_USE_TLS', 'false') === 'true';

// Default Redis TTL
export const REDIS_CACHE_TTL_SECONDS = +env.varOrDefault(
  'REDIS_CACHE_TTL_SECONDS',
  `${60 * 60 * 8}`, // 8 hours by default
);

// Whether or not to cleanup filesystem header cache files
export const ENABLE_FS_HEADER_CACHE_CLEANUP =
  env.varOrDefault('ENABLE_FS_HEADER_CACHE_CLEANUP', 'false') === 'true';

// The threshold in seconds to cleanup the filesystem contiguous data cache
export const CONTIGUOUS_DATA_CACHE_CLEANUP_THRESHOLD = env.varOrDefault(
  'CONTIGUOUS_DATA_CACHE_CLEANUP_THRESHOLD',
  '',
);

//
// Webhooks
//

// The webhook target servers
export const WEBHOOK_TARGET_SERVERS_VALUE = env.varOrUndefined(
  'WEBHOOK_TARGET_SERVERS',
);
export const WEBHOOK_TARGET_SERVERS =
  WEBHOOK_TARGET_SERVERS_VALUE !== undefined
    ? WEBHOOK_TARGET_SERVERS_VALUE.split(',')
    : [];

// The index filter to use for webhooks
export const WEBHOOK_INDEX_FILTER_STRING = canonicalize(
  JSON.parse(env.varOrDefault('WEBHOOK_INDEX_FILTER', '{"never": true}')),
);
export const WEBHOOK_INDEX_FILTER = createFilter(
  JSON.parse(WEBHOOK_INDEX_FILTER_STRING),
);

// Block filter to use for webhooks
export const WEBHOOK_BLOCK_FILTER_STRING = canonicalize(
  JSON.parse(env.varOrDefault('WEBHOOK_BLOCK_FILTER', '{"never": true}')),
);
export const WEBHOOK_BLOCK_FILTER = createFilter(
  JSON.parse(WEBHOOK_BLOCK_FILTER_STRING),
);

//
// ArNS Resolution
//

export const AR_IO_SDK_LOG_LEVEL = env.varOrDefault(
  'AR_IO_SDK_LOG_LEVEL',
  'none',
);

export const ARNS_CACHE_TYPE = env.varOrDefault('ARNS_CACHE_TYPE', 'node');

export const ARNS_CACHE_TTL_SECONDS = +env.varOrDefault(
  'ARNS_CACHE_TTL_SECONDS',
  `${60 * 60}`, // 1 hour
);

export const ARNS_RESOLVER_OVERRIDE_TTL_SECONDS_STRING = env.varOrUndefined(
  'ARNS_RESOLVER_OVERRIDE_TTL_SECONDS',
);

export const ARNS_RESOLVER_OVERRIDE_TTL_SECONDS =
  ARNS_RESOLVER_OVERRIDE_TTL_SECONDS_STRING !== undefined
    ? +ARNS_RESOLVER_OVERRIDE_TTL_SECONDS_STRING
    : undefined;

export const ARNS_CACHE_MAX_KEYS = +env.varOrDefault(
  'ARNS_CACHE_MAX_KEYS',
  '10000',
);

export const ARNS_RESOLVER_PRIORITY_ORDER = env
  .varOrDefault('ARNS_RESOLVER_PRIORITY_ORDER', 'on-demand,gateway')
  .split(',');

export const ARNS_ON_DEMAND_CIRCUIT_BREAKER_TIMEOUT_MS = +env.varOrDefault(
  'ARNS_ON_DEMAND_CIRCUIT_BREAKER_TIMEOUT_MS',
  `${5 * 1000}`, // 5 seconds
);

export const ARNS_ON_DEMAND_CIRCUIT_BREAKER_ERROR_THRESHOLD_PERCENTAGE =
  +env.varOrDefault(
    'ARNS_ON_DEMAND_CIRCUIT_BREAKER_ERROR_THRESHOLD_PERCENTAGE',
    '50',
  );

export const ARNS_ON_DEMAND_CIRCUIT_BREAKER_ROLLING_COUNT_TIMEOUT_MS =
  +env.varOrDefault(
    'ARNS_ON_DEMAND_CIRCUIT_BREAKER_ROLLING_COUNT_TIMEOUT_MS',
    `${60 * 1000}`, // 1 minute
  );

export const ARNS_ON_DEMAND_CIRCUIT_BREAKER_RESET_TIMEOUT_MS =
  +env.varOrDefault(
    'ARNS_ON_DEMAND_CIRCUIT_BREAKER_RESET_TIMEOUT_MS',
    `${5 * 60 * 1000}`, // 5 minutes
  );

export const ARNS_NAMES_CACHE_TTL_SECONDS = +env.varOrDefault(
  'ARNS_NAMES_CACHE_TTL_SECONDS',
  `${5 * 60}`, // 5 minutes
);

export const ARNS_NAME_LIST_CACHE_MISS_REFRESH_INTERVAL_SECONDS =
  +env.varOrDefault(
    'ARNS_NAME_LIST_CACHE_MISS_REFRESH_INTERVAL_SECONDS',
    `${10}`, // 10 seconds
  );

export const ARNS_NAME_LIST_CACHE_HIT_REFRESH_INTERVAL_SECONDS =
  +env.varOrDefault(
    'ARNS_NAME_LIST_CACHE_HIT_REFRESH_INTERVAL_SECONDS',
    `${60 * 60}`, // 1 hour
  );

export const ARNS_ANT_STATE_CACHE_MISS_REFRESH_INTERVAL_SECONDS =
  +env.varOrDefault(
    'ARNS_ANT_STATE_CACHE_MISS_REFRESH_INTERVAL_SECONDS',
    `${10}`, // 10 seconds
  );

export const ARNS_ANT_STATE_CACHE_HIT_REFRESH_INTERVAL_SECONDS =
  +env.varOrDefault(
    'ARNS_ANT_STATE_CACHE_HIT_REFRESH_INTERVAL_SECONDS',
    `${60 * 5}`, // 5 minutes
  );

// TODO: support multiple gateway urls
export const TRUSTED_ARNS_GATEWAY_URL = env.varOrUndefined(
  'TRUSTED_ARNS_GATEWAY_URL',
);

//
// Mempool watcher
//

export const ENABLE_MEMPOOL_WATCHER =
  env.varOrDefault('ENABLE_MEMPOOL_WATCHER', 'false') === 'true';

export const MEMPOOL_POLLING_INTERVAL_MS = +env.varOrDefault(
  'MEMPOOL_POLLING_INTERVAL_MS',
  '30000', // 30 seconds
);

//
// AWS settings
//

export const AWS_ACCESS_KEY_ID = env.varOrUndefined('AWS_ACCESS_KEY_ID');
export const AWS_SECRET_ACCESS_KEY = env.varOrUndefined(
  'AWS_SECRET_ACCESS_KEY',
);
export const AWS_REGION = env.varOrUndefined('AWS_REGION');
export const AWS_ENDPOINT = env.varOrUndefined('AWS_ENDPOINT');

export const AWS_S3_CONTIGUOUS_DATA_BUCKET = env.varOrUndefined(
  'AWS_S3_CONTIGUOUS_DATA_BUCKET',
);
export const AWS_S3_CONTIGUOUS_DATA_PREFIX = env.varOrUndefined(
  'AWS_S3_CONTIGUOUS_DATA_PREFIX',
);

//
// Development and testing
//

// Whether or not to bypass the header cache
export const SKIP_CACHE = env.varOrDefault('SKIP_CACHE', 'false') === 'true';

// The rate (0 - 1) at which to simulate request failures
export const SIMULATED_REQUEST_FAILURE_RATE = +env.varOrDefault(
  'SIMULATED_REQUEST_FAILURE_RATE',
  '0',
);

// Circuit breaker timeout for getDataParentCircuitBreaker and getDataAttributesCircuitBreaker
export const GET_DATA_CIRCUIT_BREAKER_TIMEOUT_MS = +env.varOrDefault(
  'GET_DATA_CIRCUIT_BREAKER_TIMEOUT_MS',
  '500',
);

//
// AO
//

export const AO_MU_URL = env.varOrUndefined('AO_MU_URL');
export const AO_CU_URL = env.varOrUndefined('AO_CU_URL');
export const AO_GRAPHQL_URL = env.varOrUndefined('AO_GRAPHQL_URL');
export const AO_GATEWAY_URL = env.varOrUndefined('AO_GATEWAY_URL');
